pkg/endpoint/endpoint.go

```diff
@@ -535,29 +531,30 @@ func (e *Endpoint) RunK8sCiliumEndpointSync() {
 		controller.ControllerParams{
 			RunInterval: 10 * time.Second,
 			DoFunc: func() (err error) {
-				var (
-					podName    string
-					namespace  string
-					isHealthEP = e.HasLabels(pkgLabels.LabelHealth)
-				)
-
-				switch isHealthEP {
-				case true:
-					podName = HealthCEPPrefix + node.GetName()
-					namespace = ReservedCEPNamespace
-
-				case false:
-					podName = e.GetK8sPodName()
-					if podName == "" {
-						scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s pod name")
-						return nil
-					}
 
-					namespace = e.GetK8sNamespace()
-					if namespace == "" {
-						scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s namespace")
-						return nil
-					}
+				// The health endpoint doesn't really exist in k8s and updates to it caused
+				// arbitrary errors. Disable the controller for these endpoints
+				// This is inside the controller to avoid deadlocking; calls to
+				// RunK8sCiliumEndpointSync hold ep.Mutex and that conflicts with
+				// HasLabels.
+				if isHealthEP := e.HasLabels(pkgLabels.LabelHealth); isHealthEP {
+					// use an error object so we can stop the controller. This is NOT an
```

Comment:
Instead of starting the controller, why not create a private function, `hasLabelsRLocked` which assumes the RLock is held and then call it outside of the controller? You can use this function within `HasLabels`. This solves the potential deadlock. We are trying to induce behavior that indicates the controller shouldn't be ran; we should not run the controller and add the necessary code to avoid deadlock.

Reply:
I had hoped to keep the controller create function free of the endpoint. At least, not have to hold a lock. As it turns out, we create this controller on `e.controllers` (it was once unattached to a local controller, only the global one).
With that in mind, what you're suggesting works far better. I've also labelled the function as needing the rlock.